Delayed middleware delete does not allow move operations

gsamokovarov opened the pull request:
While trying to fix #16433, we made the middleware deletions always
happen at the end. While this works for the case of deleting the
`Rack::Runtime` middleware, it makes operations like the following
misbehave.

```ruby
gem "bundler", "< 1.16"

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  middleware.insert_after ActionDispatch::Session::CookieStore, ::Rails::Rack::Logger, config.log_tags
  middleware.delete ::Rails::Rack::Logger
end

require "minitest/autorun"
require "rack/test"

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_returns_success
    get "/"
    assert last_response.ok?
  end

  private
    def app
      Rails.application
    end
end
```

In the case ☝️  the ::Rails::Rack::Logger would be deleted instead of
moved, because the order of middleware stack building execution will be:

```ruby
[:insert, ActionDispatch::Session::CookieStore, [::Rails::Rack::Logger]]
[:delete, ::Rails::Rack::Logger, [config.log_tags]]
```

This is pretty surprising and hard to reason about behaviour unless you
go spelunking into the Rails configuration code.

I have a few solutions in mind and all of them have their drawbacks.

1. Introduce a `Rails::Configuration::MiddlewareStackProxy#delete!` that
delays the deleted operations. This will make `#delete` to be executed
in order. The drawback here is backward incompatible behavior and a new
public method.

2. Just revert to the old operations. This won't allow people to delete
the `Rack::Runtime` middleware.

3. Legitimize the middleware moving with the new `#move_after` and
`#move_before` methods. This does not break any backward
compatibility but includes 2 new methods to the middleware stack.

I have implemented `3.` in this pull request.

---

I'm reopening this after https://github.com/rails/rails/pull/31555.

Edouard-chin reviewed (commented):
I think this will cause issues if you try to move a middleware that has arguments, it will remove them

```ruby
@stack.use(FirstMiddleware)
@stack.use(SecondMiddleWare, "hello")
@stack.move_before(FirstMiddleware, SecondMiddleware)
@stack.first.args == ["hello"]
```

We should add a test similar to this one https://github.com/rails/rails/pull/36769/files#diff-2fb7d57901c3d7c8f36753118084f0e3R96-R101

gsamokovarov commented:
Thanks @Edouard-chin, will fix this!

rafaelfranca commented:
Did you fix the arguments problem? I see no test for it.

gsamokovarov commented:
Working on it at the moment, will ping you when ready.

gsamokovarov committed:
Delayed middleware delete does not allow move operations

While trying to fix #16433, we made the middleware deletions always
happen at the end. While this works for the case of deleting the
Rack::Runtime middleware, it makes operations like the following
misbehave.

```ruby
gem "bundler", "< 1.16"

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  middleware.insert_after ActionDispatch::Session::CookieStore, ::Rails::Rack::Logger, config.log_tags
  middleware.delete ::Rails::Rack::Logger
end

require "minitest/autorun"
require "rack/test"

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_returns_success
    get "/"
    assert last_response.ok?
  end

  private
    def app
      Rails.application
    end
end
```

In the case ☝️  the ::Rails::Rack::Logger would be deleted instead of
moved, because the order of middleware stack building execution will be:

```ruby
[:insert, ActionDispatch::Session::CookieStore, [::Rails::Rack::Logger]]
[:delete, ::Rails::Rack::Logger, [config.log_tags]]
```

This is pretty surprising and hard to reason about behaviour, unless you
go spelunking into the Rails configuration code.

I have a few solutions in mind and all of them have their drawbacks.

1. Introduce a `Rails::Configuration::MiddlewareStackProxy#delete!` that
delays the deleted operations. This will make `#delete` to be executed
in order. The drawback here is backwards incompatible behavior and a new
public method.

2. Just revert to the old operations. This won't allow people to delete
the `Rack::Runtime` middleware.

3. Legitimize the middleware moving with the new `#move_after` and
`#move_before` methods. This does not breaks any backwards
compatibility, but includes 2 new methods to the middleware stack.

I have implemented `3.` in this pull request.

Happy holidays! 🎄

gsamokovarov commented:
This should be ready. @Edouard-chin I covered your case and cleaned up the interface of `#move` and `#move_after` without the argument delegation clutter.

Edouard-chin reviewed (approved):
🥇

kamipo commented on railties/lib/rails/configuration.rb:
```diff
@@ -65,6 +74,19 @@ def use(*args, &block)
       def delete(*args, &block)
         @delete_operations << -> middleware { middleware.send(__method__, *args, &block) }
       end
+      ruby2_keywords(:delete) if respond_to?(:ruby2_keywords, true)
```
These three `ruby2_keywords` are not necessary because `delete`, `move_before`, and `move_after` doesn't take keyword arguments.

gsamokovarov replied:
Thanks, I've removed the `ruby2_keywords` from the Action Pack code yesterday, but did not take the proxy into account. I have removed them from here as well!
